### DISCUSSION TITLE
New experimental docker image

### DISCUSSION THREAD

**Discussion by jbtronics** [Part-DB/Part-DB-server#562] [Announcements]
# New experimental docker image

There now is a new (still experimental) docker image of Part-DB: [partdborg/part-db](https://hub.docker.com/r/partdborg/part-db)

Unlike the jbtronics/part-db1 image, which uses a classic apache and php-fpm setup, the new image is based on [FrankenPHP](https://frankenphp.dev/) a PHP application server based on [Caddy](https://caddyserver.com/).

The main advantage is that FrankenPHP supports a worker mode for PHP execution, where the initialized symfony kernel can be reused for multiple requests. On the classic PHP request model on apache, nginx, etc. the symfony kernel has to be loaded and reinitialized again on every request, which slow everything down.

So with the new FrankenPHP based docker image, everything should be much faster. First experiments showed something like 10 ms on the FrankenPHP image compared to 50 ms with the classical approach on pages with few database operations. So a pretty good improvement of request time.

Another advantage is that Caddy (the webserver FrankenPHP is based on) has very easy TLS handling, somethin which was not possible so easy with the old images. Based on the easy configuration Caddy can automatically create self-signed TLS certificates, use existing certificates or automatically request TLS certificates from a ACME provider like LetsEncrypt. That way setting up a secure access to Part-DB should become much easier.

It is planned that the new image should replace the old one in the future as the default image. However as it changes some things and especially the worker mode was never tested with Part-DB before, it is currently only experimental.

There are some changes to the current docker interface, so it is required to use some new docker compose configuration.
If you wanna try it out create a file named `compose.yaml` with the following content and modify it to your needs:

```docker-compose
services:
  php:
    image: partdborg/part-db:edge
    restart: unless-stopped
    environment:
      SERVER_NAME: ${SERVER_NAME:-localhost}, php:80
      MERCURE_PUBLISHER_JWT_KEY: ${CADDY_MERCURE_JWT_SECRET:-!ChangeThisMercureHubJWTSecretKey!}
      MERCURE_SUBSCRIBER_JWT_KEY: ${CADDY_MERCURE_JWT_SECRET:-!ChangeThisMercureHubJWTSecretKey!}
      TRUSTED_PROXIES: ${TRUSTED_PROXIES:-[127.0.0.0/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16](https://127.0.0.0/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16)}
      TRUSTED_HOSTS: ^${SERVER_NAME:-example\.com|localhost}|php$$
      MERCURE_URL: ${CADDY_MERCURE_URL:-http://php/.well-known/mercure}
      MERCURE_PUBLIC_URL: https://${SERVER_NAME:-localhost}/.well-known/mercure
      MERCURE_JWT_SECRET: ${CADDY_MERCURE_JWT_SECRET:-!ChangeThisMercureHubJWTSecretKey!}

     # Add your env options here like normally
     DATABASE_URL: mysql://root:@host.docker.internal:3306/partdb?serverVersion=10.6.5-mariadb    

    volumes:
      - caddy_data:/data
      - caddy_config:/config
      - ./uploads:/var/www/html/uploads
      - ./public_media:/var/www/html/public/media
      - ./db:/var/www/html/var/db
    ports:
      # HTTP
      - target: 80
        published: 4080
        protocol: tcp
      # HTTPS
      - target: 443
        published: 4443
        protocol: tcp
      # HTTP/3
      - target: 443
        published: 4443
        protocol: udp
        
    extra_hosts:
      # Ensure that host.docker.internal is correctly defined on Linux
      - host.docker.internal:host-gateway
        
        
volumes:
  caddy_data:
  caddy_config:
```

This file might change in the future. You can start Part-DB then using  `docker compose up` (not `docker-compose`)

Also beware that this is based on the development versions of Part-DB and automatically updates the database. So this version may contain errors and you should regulary do database backups!

**Comment by ishioni** [comment]
I tried using this in a kubernetes environment, but it doesn't want to play nice - caddy keeps trying to generate a TLS certificate (which is unnecessary given that people run ingress controllers there), and once i disable that with CADDY_GLOBAL_OPTIONS: auto_https off it renders an empty page on port 80. Any hints?

**Comment by Applearcher80** [comment]
Are there instructions for updating from 17.1.0 docker image to 17.1.1?

**Comment by salfter** [comment]
I'm already running Caddy (caddy-docker-proxy, to be more precise) to manage access to multiple services on the server that hosts my PartDB instance.  Several of these (including some bigger ones, like Nextcloud) run php-fpm without a webserver; Caddy is then configured in the application's docker-compose.yml to talk to the service.  It'd be nice to have PartDB packaged similarly, whether using php-fpm or a faster alternative that presents a compatible interface to whatever reverse proxy server you're already running.

**Reply by jbtronics** [reply]
FrankenPHP (which this image use) is more than just a simple caddy webserver. The relevant thing here is the worker mode, which allows the initialized application kernel to handle multiple requests, which is just not possible with the classic PHP modes (and requires some abstraction inside the application). That will save multiple milliseconds of initialization time every request, while the overhead due to the additional layer of web server will be neglectable...
